cameo/flux_analysis/analysis.py

```diff
@@ -499,14 +500,62 @@ def _get_c_source_reaction(model):
        The medium reaction with highest input carbon flux
     """
     try:
-        model.optimize()
-    except (Infeasible, AssertionError):
+        model.solver.optimize()
+        assert_optimal(model)
+    except OptimizationError:
         return None
-    model_reactions = model.reactions.get_by_any(list(model.medium))
-    source_reactions = [(reaction, reaction.flux * n_carbon(reaction)) for reaction in model_reactions if
-                        reaction.flux < 0]
-    sorted_sources = sorted(source_reactions, key=lambda reaction_tuple: reaction_tuple[1])
-    return sorted_sources[0][0]
+    medium_reactions = model.reactions.get_by_any(list(model.medium))
+    medium_reactions_fluxes = [(rxn, total_carbon_flux(rxn, consumption=True)) for rxn in medium_reactions]
+    source_reactions = [(rxn, c_flux) for rxn, c_flux in medium_reactions_fluxes if c_flux > 0]
+    sorted_sources = sorted(source_reactions, key=itemgetter(1))
```

Comment:
If only the maximum element is of interest, rather than sorting the container, I'd use `max(source_reactions, key=itemgetter(1))`. Didn't think about that before, sorry.